ARExperiment/ARExperiment/RecognizeObjectsViewController.swift

```diff
@@ -92,78 +87,172 @@ class RecognizeObjectsViewController: UIViewController, ARSCNViewDelegate, ARExp
-            return
+
+        guard let prediction = predictionForTinyYOLO(topObservation: topObservation),
+            let objectBounds = boundingBoxRectForTinyYOLO(prediction: prediction) else {
+                showStartButtonIfError()
+                print("Tiny YOLO failed finding top prediction" )
+                return nil
         }
-        showOnMainThread(boundingBoxes)
+        return ObjectPrediction(name: objectNameForTinyYOLO(prediction: prediction), bounds: objectBounds)
     }
-    
-    
-    //MARK: Prediction
-    
-    private func showOnMainThread(_ boundingBoxes: [YOLO.Prediction]) {
-        let prominentBox = boundingBoxes.sorted{ $0.score > $1.score}.first
-        self.semaphore.signal()
-        
+
+    private func processPredictionFromVision(results: [Any]?) -> ObjectPrediction? {
+        guard #available(iOS 12.0, *) else {
+            usingTinyModel = true
+           return nil
+        }
+        guard let observations = results as? [VNRecognizedObjectObservation],
+            let topObservation = observations.first,
+            let topLabelObservation = topObservation.labels.first else {
+                print("Vision failed finding prediction" )
+                showStartButtonIfError()
+                return nil
+        }
+        let objectRect = VNImageRectForNormalizedRect(topObservation.boundingBox,
+                                                  Int(currentSceneFrame.width),
+                                                  Int(currentSceneFrame.height))
+        return ObjectPrediction(name: topLabelObservation.identifier, bounds: objectRect)
+    }
+
+    private func predictionForTinyYOLO(topObservation: MLMultiArray) -> YOLO.Prediction? {
+        let boundingBoxes = yolo.computeBoundingBoxes(features: topObservation)
+        let prominentBox = boundingBoxes.sorted{ $0.score > $1.score}
+        return prominentBox.first
+    }
+
+    private func boundingBoxRectForTinyYOLO(prediction: YOLO.Prediction) -> CGRect? {
+        let viewRect = CGRect(x: 0, y: 0, width: Int(currentSceneFrame.width), height: Int(currentSceneFrame.height))
+        return yolo.scaleImageForCameraOutput(predictionRect: prediction.rect, viewRect: viewRect)
+    }
+
+    private func objectNameForTinyYOLO(prediction: YOLO.Prediction) -> String {
+        return labels[prediction.classIndex]
+    }
+
+    //MARK: Add Model to the scene
+
+    private func addBoxOnMainThread(_ prediction: ObjectPrediction?) {
+        guard let prediction = prediction,
+            let hitPoint = findHitPointFor(prediction.bounds) else {
+            showStartButtonIfError()
+            print("failed finding hit point or prediction")
+            return
+        }
         DispatchQueue.main.async { [weak self] in
-            if let prominentBox = prominentBox {
-                self?.show(prediction: prominentBox)
+            if self?.usingAnchors ?? false {
```

Comment:
should we maybe strongify the self reference back to avoid all this not needed optionalities? 
Can't we maybe assume that self is always going to exist here and use `unowned`?

Reply:
remember unowned is basically a force unwrap, often not ideal.

Perhaps you can `guard let strongSelf = self` and use that instead

Reply:
I'm getting tired of this weakify/strongify little dance.
(BTW: Not necessary if you have only one line (no atomic guarantee needed), so you can keep the optionality).

And I like when you *know* that `unowned` will work and do it. Most of the time there is no cycle.

But sure when in doubt (or don't have time) use that little dance... I wish there was a rule of thumb easy to remember to know when `unowned` will always work!

On TheSun (with @gbasile) we tried hard at the end to use `unowned` a lot more.

Reply:
@malaba If you're getting tired you should move to developing on a garbage collected system :P 

the problem is the same as a force unwrap, you're making an assumption that won't necessarily always be correct. 

I do believe however that Swift's memory management is incredibly painful. What I tend to do is either make this extension:

```swift
extension Optional {
  func ifNotNil(_ action: (Wrapped) -> Void) {
     ...
  }
}
```

I also tried out this last night:

```swift
func byWeakly<Reference: AnyObject>(referencing reference: Reference, inClosure action: @escaping (Reference) -> Void) -> () -> Void {
    weak var weakRef = reference
    return {
        if let strongRef = weakRef {
            action(strongRef)
        }
    }
}

func byWeakly<Input, Reference: AnyObject>(referencing reference: Reference, inClosure action: @escaping (Reference, Input) -> Void) -> (Input) -> Void {
    weak var weakRef = reference
    return { input in
        if let strongRef = weakRef {
            action(strongRef, input)
        }
    }
}
```

which you can use like this where `doSomethingLongLived ` is an escaping function:

```swift
doSomethingLongLived(byWeakly(referencing: self) { strongSelf, int in
    strongSelf.value = strongSelf.value + int
})
```